proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+       scPtrOptDonated <- threadGet tcbSchedContext callerPtr;
+       tptrOpt <- getReplyTCB replyPtr;
+       haskell_assert (tptrOpt = Nothing) [];
+       scPtrOptCallee <- threadGet tcbSchedContext calleePtr;
+       replyPush' callerPtr calleePtr replyPtr canDonate scPtrOptDonated scPtrOptCallee;
+       when (scPtrOptDonated \<noteq> Nothing \<and> (if scPtrOptCallee \<noteq> Nothing then False else canDonate)) $
+           schedContextDonate (fromJust scPtrOptDonated) calleePtr od)
+   (replyPush callerPtr calleePtr replyPtr canDonate)"
+  unfolding replyPush_def replyPush'_def
+  apply (clarsimp simp only: K_bind_def bind_assoc)
+  apply (rule monadic_rewrite_gen_asm[simplified pred_conj_def]
+              monadic_rewrite_bind_tail[OF _ return_sp]
+              monadic_rewrite_bind_tail[OF _ hoare_post_taut])+
+  apply (case_tac "(scPtrOptDonated \<noteq> Nothing \<and> x)")
+   apply (simp add: bind_assoc monadic_rewrite_refl split: if_splits)
+  apply (simp only: when_False fun_app_def unit_abs_eta_conv bind_return monadic_rewrite_refl)
+  done
+
+crunches replyPush'
+  for tcb_at'[wp]: "tcb_at' t"
+  and ex_nonz_cap_to'[wp]: "ex_nonz_cap_to' p"
+  and valid_queues[wp]: "valid_queues"
+  (wp: crunch_wps hoare_vcg_all_lift simp: crunch_simps)
+
+lemma replyPush'_valid_objs'[wp]:
+  "replyPush' callerPtr calleePtr replyPtr canDonate scPtrOptDonated scPtrOptCallee \<lbrace>valid_objs'\<rbrace>"
+  apply (solves wp | simp (no_asm_use) add: replyPush'_def split del: if_split cong: conj_cong |
+         wp hoare_when_wp haskell_assert_wp hoare_vcg_if_lift hoare_vcg_all_lift
+            hoare_vcg_disj_lift hoare_vcg_imp_lift' set_sc'.valid_objs'
+            set_reply'.valid_objs' setReply_obj_at' setReplyTCB_obj_at')+
+  apply (clarsimp simp: valid_tcb_state'_def ko_wp_at'_def obj_at'_def)
+  apply (fastforce elim: valid_objsE'[where x="the scPtrOptDonated"]
+                   simp: valid_obj'_def valid_sched_context'_def valid_sched_context_size'_def
+                         valid_reply'_def valid_bound_obj'_def objBits_def objBitsKO_def
+                         obj_at'_def projectKO_eq projectKO_reply projectKO_tcb projectKO_sc)
+  done
+
+lemma replyPush_valid_objs'[wp]:
+  "replyPush callerPtr calleePtr replyPtr canDonate \<lbrace>valid_objs'\<rbrace>"
+  by (rule monadic_rewrite_refine_valid[OF monadic_rewrite_replyPush, where P''=\<top>, simplified])
+     (wpsimp wp: schedContextDonate_valid_objs' threadGet_wp hoare_vcg_if_lift hoare_vcg_imp_lift
+           simp: getReplyTCB_def obj_at'_def split_del: if_split)
+
+lemma replyPush_sch_act_wf:
+  "\<lbrace>\<lambda>s. sch_act_wf (ksSchedulerAction s) s \<and> sch_act_not callerPtr s\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_ s. sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  unfolding replyPush_def getReplyTCB_def
+  by (wpsimp wp: sts_sch_act' hoare_vcg_all_lift hoare_vcg_if_lift hoare_drop_imps)
+
+lemma replyPush_sym_refs_helper:
```

Comment:
Could you put `list_refs_of_replies'` in the name of this one and the following lemma to distinguish it from regular old `sym_refs (state_refs_of' s)`?